openedx/core/djangoapps/session_inactivity_timeout/middleware.py

```diff
@@ -34,27 +39,73 @@ def process_request(self, request):
+            # .. setting_description: How often to allow a full session save (in seconds).
+            #   This controls how frequently the session ID might change.
+            #   A user could be inactive for almost SESSION_ACTIVITY_SAVE_DELAY_SECONDS but since their session
+            #   isn't being saved during that time, their last activity timestamp isn't being updated.
+            #   When they hit the inactivity timeout, it will be based on the last saved activity time.
+            #   So the effective timeout could be as short as:
+            #   SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_ACTIVITY_SAVE_DELAY_SECONDS.
+            #   This means users might be logged out earlier than expected in some edge cases.
+            # .. setting_warning:  Must be smaller than SESSION_INACTIVITY_TIMEOUT_IN_SECONDS.
+            frequency_time_in_seconds = getattr(settings, "SESSION_ACTIVITY_SAVE_DELAY_SECONDS", 900)
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
 
             # have we stored a 'last visited' in session? NOTE: first time access after login
             # this key will not be present in the session data
-            if last_touch:
-                # compute the delta since last time user came to the server
-                time_since_last_activity = utc_now - last_touch
-
-                # did we exceed the timeout limit?
-                if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
-                    # yes? Then log the user out
-                    del request.session[LAST_TOUCH_KEYNAME]
-                    auth.logout(request)
-                    return
-
-            request.session[LAST_TOUCH_KEYNAME] = utc_now
+            if last_touch_str:
+                try:
+                    last_touch = datetime.fromisoformat(last_touch_str)
+                    time_since_last_activity = current_time - last_touch
+
+                    has_exceeded_timeout_limit = time_since_last_activity > timedelta(seconds=timeout_in_seconds)
+                    if has_exceeded_timeout_limit:
+                        del request.session[LAST_TOUCH_KEYNAME]
+                        auth.logout(request)
+                        return
+                except (ValueError, TypeError) as e:
+                    # If parsing fails, log warning and then treat as if no timestamp exists
+                    log.warning("Parsing last touch time failed: %s", e)
+                    # .. custom_attribute_name: session_inactivity.last_touch_error
+                    # .. custom_attribute_description: Boolean. True if parsing the last activity timestamp failed for this request, indicating a session data error.
+                    monitoring_utils.set_custom_attribute('session_inactivity.last_touch_error', True)
```

Comment:
I didn't mean all custom attributes needed to be booleans. This was a perfectly good use of a string. :)
```suggestion
                    monitoring_utils.set_custom_attribute('session_inactivity.last_touch_error', str(e))
```

Reply:
> I didn't mean all custom attributes needed to be booleans. This was a perfectly good use of a string. :)

Is this required to add exception in monitoring. As except block handles robust error handling for cases where the stored timestamp data is corrupted, malformed, or otherwise unparseable. 
If you are looking where someone is looking into Datadog and needs to understand details of the error then I can update with changes recommended by you.

Reply:
Since you are handling the error, DD will not get it and no one would know about it. My suggestion would provide some of the error details, rather than True.

Additionally, and even better, it is probably worth-while using [record_exception](https://github.com/search?q=org%3Aopenedx%20record_exception&type=code), to sent it to DD. Note: this only needs to be done where an exception is being handled **and** it is still unexpected and warrants looking into.

Reply:
> Since you are handling the error, DD will not get it and no one would know about it. My suggestion would provide some of the error details, rather than True.
> 
> Additionally, and even better, it is probably worth-while using [record_exception](https://github.com/search?q=org%3Aopenedx%20record_exception&type=code), to sent it to DD. Note: this only needs to be done where an exception is being handled **and** it is still unexpected and warrants looking into.

Thanks, will add error details for this.

Will ignore record_exeception for this as it can be triggered for corrupted, malformed, or otherwise unparseable timestamp which does not warrant us to look into that.

Let me know your thoughts on this. If you feel otherwise I can add this.

Reply:
We'll use `record_exception` to know if and when a bug gets introduced, because this really shouldn't fire. You can leave the redundant custom attribute or not for the error as you see fit.

Reply:
added record_exception()